File: .github/workflows/robot/internal/bot/label_test.go
/*
Copyright 2021 Gravitational, Inc.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

package bot

import (
	"context"
	"testing"

	"github.com/gravitational/teleport/.github/workflows/robot/internal/env"

	"github.com/stretchr/testify/require"
)

// TestLabel checks that labels are correctly applied to a Pull Request.
func TestLabel(t *testing.T) {
	tests := []struct {
		desc   string
		branch string
		files  []string
		labels []string
	}{
		{
			desc:   "code-only",
			branch: "foo",
			files: []string{
				"file.go",
				"examples/README.md",
			},
			labels: []string{},
		},
		{
			desc:   "docs",
			branch: "foo",
			files: []string{
				"docs/docs.md",
			},
			labels: []string{"documentation"},
		},
		{
			desc:   "helm",
			branch: "foo",
			files: []string{
				"examples/chart/index.html",
			},
			labels: []string{"helm"},
		},
		{
			desc:   "docs-and-helm",
			branch: "foo",
			files: []string{
				"docs/docs.md",
				"examples/chart/index.html",
			},
			labels: []string{"documentation", "helm"},
		},
		{
			desc:   "docs-and-backport",
			branch: "branch/foo",
			files: []string{
				"docs/docs.md",
			},
			labels: []string{"backport", "documentation"},
		},
	}
	for _, test := range tests {
		t.Run(test.desc, func(t *testing.T) {
			b := &Bot{
				c: &Config{
					Environment: &env.Environment{
						Organization: "foo",
						Repository:   "bar",
						Number:       0,
						UnsafeBranch: test.branch,
					},
					GitHub: &fakeGithub{
						test.files,
					},
				},
			}
			labels, err := b.labels(context.Background())
			require.NoError(t, err)
			require.ElementsMatch(t, labels, test.labels)
		})
	}
}

File: web/packages/design/src/Label/Label.test.js
/**
 * Copyright 2020 Gravitational, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React from 'react';

import { render, theme } from 'design/utils/testing';

import Label, { Primary, Secondary, Warning, Danger } from './Label';

describe('design/Label', () => {
  const colors = [
    theme.colors.brand,
    theme.colors.levels.sunkenSecondary,
    theme.colors.warning.main,
    theme.colors.danger,
  ];

  test.each`
    kind                  | Component    | expected
    ${'default: primary'} | ${Label}     | ${colors[0]}
    ${'primary'}          | ${Primary}   | ${colors[0]}
    ${'secondary'}        | ${Secondary} | ${colors[1]}
    ${'warning'}          | ${Warning}   | ${colors[2]}
    ${'danger'}           | ${Danger}    | ${colors[3]}
  `('component renders $kind label', ({ Component, expected }) => {
    const { container } = render(<Component />);
    expect(container.firstChild).toHaveStyle({
      'background-color': expected,
    });
  });
});
